tests/test.py

```diff
@@ -4529,310 +4242,61 @@ class TestZoneIDDecryptionWithVaultMasterKeyLoader(HashiCorpVaultMasterKeyLoader
+            ca_bytes = ca_file.read()
+            key_bytes = key_file.read()
+            cert_bytes = cert_file.read()
+        tls_credentials = grpc.ssl_channel_credentials(ca_bytes, key_bytes, cert_bytes)
+        return grpc.secure_channel('localhost:{}'.format(port), tls_credentials)
 
-    def setUp(self):
-        self.checkSkip()
+    def get_http_default_kwargs(self):
+        '''setup requests to use client's certificates'''
+        kwargs = super().get_http_default_kwargs()
+        kwargs['verify'] = TEST_TLS_CA
+        # https://requests.readthedocs.io/en/master/user/advanced/#client-side-certificates
+        # first crt, second key
+        kwargs['cert'] = (TEST_TLS_CLIENT_CERT, TEST_TLS_CLIENT_KEY)
+        return kwargs
 
     def apiEncryptionTest(self, request_func, use_http=False, use_grpc=False):
         # one is set
         self.assertTrue(use_http or use_grpc)
         # two is not acceptable
         self.assertFalse(use_http and use_grpc)
         translator_port = 3456
-        connector_port = 12345
-        connector_port2 = connector_port+1
-        client_id = "keypair1"
-        data = get_pregenerated_random_data().encode('ascii')
-        client_id_private_key = read_storage_private_key(KEYS_FOLDER.name, 'keypair1')
-        zone = zones[0]
-        zone_private_key = read_zone_private_key(KEYS_FOLDER.name, zone['id'])
-        connection_string = 'tcp://localhost:{}'.format(translator_port)
-        translator_kwargs = {
-            'incoming_connection_http_string': connection_string if use_http else '',
-            # turn off grpc to avoid check connection to it without acra-connector
-            'incoming_connection_grpc_string': connection_string if use_grpc else '',}
-
-        correct_client_id = 'keypair1'
-        incorrect_client_id = 'keypair2'
-        with ProcessContextManager(self.fork_translator(translator_kwargs)):
-            with ProcessContextManager(self.fork_connector_for_translator(connector_port, translator_port, client_id)):
-                response = request_func(connector_port, correct_client_id, None, data)
-                decrypted = deserialize_and_decrypt_acrastruct(response, client_id_private_key, 'keypair1')
-                self.assertEqual(data, decrypted)
-                # test with correct zone id
-                print("encrypt with zone {}".format(zone['id']))
-                response = request_func(
-                    connector_port, client_id, zone['id'], data)
-                print("decrypt with zone {}".format(zone['id']))
-                decrypted = deserialize_and_decrypt_acrastruct(response, zone_private_key, zone_id=zone['id'].encode('ascii'))
+        key_folder = tempfile.TemporaryDirectory()
+        try:
+            client_id = self.get_valid_certificate_identifier()
+            self.assertEqual(create_client_keypair(name=client_id, keys_dir=key_folder.name), 0)
```

Comment:
plus better to generate keys from certificate instead of extracted client_id, as it will do users) and we have function for that

Reply:
Actually we do generate keys from cert but not extract, probably we should change the actual function name.
Here is the source of get_valid_certificate_identifier:

```
    def get_valid_certificate_identifier(self, certificate_file=None):
        if certificate_file is None:
            certificate_file = TEST_TLS_CLIENT_CERT
        # converted data from certificate "CN=Test leaf certificate (acra-writer),OU=IT,O=Global Security,L=London,ST=London,C=GB"
        with open(certificate_file, 'rb') as f:
            pem = f.read()
        cert = x509.load_pem_x509_certificate(pem, default_backend())
        sha512 = hashlib.sha512()
        sha512.update(cert.subject.rfc4514_string().encode('utf-8'))
        return sha512.hexdigest().lower()
```

Reply:
hm... yes, we should change it)) because this logic may work differently from how it works in golang code (that is main reason why we provided `extract-client-id` command for `acra-keys`). So... it I left a time bomb *(

But my original question was related to flow `get client_id` and `call create_client_keypair` instead of `create_client_keypair_from_certificate`. Why we get client_id from somewhere and pass it to `create_client_keypair` instead of calling `acra-keymaker` passing path to certificate?